library/std/src/net/ip.rs

```diff
@@ -1056,27 +1070,12 @@ impl Ipv6Addr {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     #[rustc_const_stable(feature = "const_ipv6", since = "1.32.0")]
+    #[allow_internal_unstable(const_fn_union)]
     pub const fn new(a: u16, b: u16, c: u16, d: u16, e: u16, f: u16, g: u16, h: u16) -> Ipv6Addr {
         Ipv6Addr {
             inner: c::in6_addr {
-                s6_addr: [
-                    (a >> 8) as u8,
-                    a as u8,
-                    (b >> 8) as u8,
-                    b as u8,
-                    (c >> 8) as u8,
-                    c as u8,
-                    (d >> 8) as u8,
-                    d as u8,
-                    (e >> 8) as u8,
-                    e as u8,
-                    (f >> 8) as u8,
-                    f as u8,
-                    (g >> 8) as u8,
-                    g as u8,
-                    (h >> 8) as u8,
-                    h as u8,
-                ],
+                // SAFETY: Each element of `s6_addr16` is big edian.
+                s6_addr: unsafe { in6_union { s6_addr16: [a, b, c, d, e, f, g, h] }.s6_addr },
```

Comment:
The assumption here is wrong about arguments: (`a`, `b`, ...).
Those are not guaranteed to be Big Endian.

Reply:
Right, the parameters are native-endian, and the output from `segments` is native too.

Maybe you could use `[a.to_be_bytes(), b.to_be_bytes, ...]` to get `[[u8; 2]; 8]` instead, and then cast that into `s6_addr`. But once we're dealing with byte swaps, your godbolt advantage may disappear, so I think this will only be a matter of code style.

Reply:
Or `[a.to_be(), b.to_be(), ...]` to get a `[u16; 8]` directly that is big endian.

Reply:
This actually does lead to better instructions: https://rust.godbolt.org/z/dv4Pje